ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -706,53 +651,63 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+    -- discussed above they become available.
     rollForward :: MkPipelineDecision
                 -> Nat n
                 -> Header blk
                 -> Their (Tip blk)
-                -> Stateful m blk
+                -> StatefulConsensus m blk
                      (KnownIntersectionState blk)
                      (ClientPipelinedStIdle n)
     rollForward mkPipelineDecision n hdr theirTip
               = Stateful $ \kis -> traceException $ do
-      -- Reject the block if invalid
-      let hdrHash  = headerHash hdr
+      let hdrSlot  = blockSlot   hdr
           hdrPoint = headerPoint hdr
+      -- Reject the block if invalid
       isInvalidBlock <- atomically $ forgetFingerprint <$> getIsInvalidBlock
-      whenJust (isInvalidBlock hdrHash) $ \reason ->
+      whenJust (isInvalidBlock (headerHash  hdr)) $ \reason ->
         disconnect $ InvalidBlock hdrPoint reason
 
       -- Get the ledger view required to validate the header
-      -- NOTE: This will block if we are too far behind.
-      intersectCheck <- atomically $ do
-        -- Before obtaining a 'LedgerView', we must find the most recent
-        -- intersection with the current chain. Note that this is cheap when
-        -- the chain and candidate haven't changed.
-        mKis' <- intersectsWithCurrentChain kis
-        case mKis' of
-          Nothing -> return NoLongerIntersects
-          Just kis'@KnownIntersectionState { mostRecentIntersection } -> do
-            -- We're calling 'ledgerViewForecastAt' in the same STM transaction
-            -- as 'intersectsWithCurrentChain'. This guarantees the former's
-            -- precondition: the intersection is within the last @k@ blocks of
-            -- the current chain.
-            forecast <-
-              maybe
-                (error $
-                   "intersection not within last k blocks: " <> show mostRecentIntersection)
-                (ledgerViewForecastAt (configLedger cfg) . ledgerState)
-                <$> getPastLedger mostRecentIntersection
-
-            case runExcept $ forecastFor forecast (blockSlot hdr) of
-              -- The header is too far ahead of the intersection point with our
-              -- current chain. We have to wait until our chain and the
-              -- intersection have advanced far enough. This will wait on
-              -- changes to the current chain via the call to
-              -- 'intersectsWithCurrentChain' befoer it.
-              Left OutsideForecastRange{} ->
-                retry
-              Right ledgerView ->
-                return $ Intersects kis' ledgerView
+      intersectCheck <- join $ atomically $ do
+        rechk <- stillIntersectsWithCurrentChain cfg cdbView kis
+        case recheckToMaybe kis rechk of
+          Nothing   -> return $ return NoLongerIntersects
+          Just kis' ->
+              continueWithState kis'
+                $ checkDensityOrContinue cfg cdbView hdrSlot
```

Comment:
Rename to `checkDensityOrContinue` to `tryToGetLedgerView` or something along those lines.

Reply:
Let's bake the `lowDensityHandler` continuation into `checkDensityOrContinue`. The current call graph does that, just avoiding passing unchanged parameters at the moment. Clarity of static calls is more preferable.